README
See the manual page macfanctld (1) for more information.


debian/changelog
macfanctld (0.7-0) raring; urgency=low

  * Bump version
  * Bump Standards-Version (no change)
  * Use fan*_output node to control fans speed
  * Code styling and cleanup
    (Sorry Mikael, I am obsessive about vertical space loss and white spaces)
  * New defaults in config

 -- Eugene San (eugenesan) <eugenesan@gmail.com>  Fri, 07 Jun 2013 10:20:11 +0300

macfanctld (0.6-0) raring; urgency=low

  * Updated for Raring.
  * Removed bogus patches.
  * Sanetize versioning.

 -- Eugene San (eugenesan) <eugenesan@gmail.com>  Thu, 18 Apr 2013 01:57:27 +0300

macfanctld (0.6~mactel1ubuntu3~raring1) raring; urgency=low

  * Update for Raring.

 -- Mikael Strom <mikael@sesamiq.com>  Fri, 26 Apr 2013 21:58:40 +0800

macfanctld (0.6~mactel1ubuntu3~quantal1) quantal; urgency=low

  * Updated for Quantal.

 -- Mikael Strom <mikael@sesamiq.com>  Mon, 22 Oct 2012 23:49:02 +0800

macfanctld (0.6~mactel1ubuntu3~quantal) precise; urgency=low

  * Updated for Quantal.

 -- Mikael Strom <mikael@sesamiq.com>  Sun, 22 Oct 2012 14:54:02 +0800

macfanctld (0.6~mactel1ubuntu3~precise) precise; urgency=low

  * Updated for Precise.

 -- Mikael Strom <mikael@sesamiq.com>  Sun, 22 Apr 2012 12:12:02 +0800

macfanctld (0.6~mactel1ubuntu2) oneiric; urgency=low

  * Decreased CPU usage by caching sensor file names and replacing 
    fopen()/fclose() with open()/close().

 -- Mikael Strom <mikael@sesamiq.com>  Sat, 21 Apr 2012 17:16:42 +0800

macfanctld (0.6~mactel1ubuntu1) oneiric; urgency=low

  * updated for Oneiric

 -- Mikael Strom <mikael@sesamiq.com>  Sat, 21 Apr 2012 14:45:41 +0800

macfanctld (0.6~mactel1) natty; urgency=low

  * Use PnP rather than hardcoding resources and devices: 
    locate applesmc via /sys/class/hwmon and check that device/name 
    is "applesmc" as suggested by Julien BLACHE. 

 -- Mikael Strom <mikael@sesamiq.com>  Wed, 19 Aug 2011 01:31:36 +0800

macfanctld (0.5~mactel1) natty; urgency=low

  * Corrected comment in macfanctl.conf

 -- Mikael Strom <mikael@sesamiq.com>  Wed, 18 Aug 2011 01:31:36 +0800

macfanctld (0.4~mactel1~karmic) karmic; urgency=low

  * Patched using Henrik Rydbergs "Off by one" patch
  * Added 'exclude' option to config file. Enables specific sensors to be excluded.

 -- Mikael Strom <mikael@sesamiq.com>  Sat, 06 Nov 2010 17:58:59 +0800

macfanctld (0.3~mactel1~maverick) maverick; urgency=low

  * Fixed empty line bug in log. Produced an empty line ecery 5 seconds.

 -- Mikael Strom <mikael@sesamiq.com>  Thu, 02 Sep 2010 03:22:15 +0800

macfanctld (0.2~mactel1~karmic) karmic; urgency=low

  * Moved to mactel repo.

 -- Mikael Strom <mikael@sesamiq.com>  Wed, 01 Sep 2010 22:24:26 +0800

macfanctld (0.2) lucid; urgency=low

  * Fixed bug where only fan 1 got updated.
  * Changed to use sensors TC0P/TG0P instead of TC0D/TC0D that was to noisy.
  * Removed the modes. Now uses TC0P/TG0P and average concurrently.
  * Replaced README with a new man page, macfanctld.1

 -- Mikael Strom <mikael@sesamiq.com>  Sun, 29 Aug 2010 04:09:51 +0800

macfanctld (0.1) lucid; urgency=low

  * Updated to GPL 3 in source files.
  * Updated README, and added to the deb-package.
  * Clean up the code, monir bugfixes.
  * Added code to handle only 1 fan.
  * Removed command line switches -a -d and -s. Moved this settings
    to the config file.
  * Fixed so that HUP signal re-reads config properly.
  
 -- Mikael Strom <mikael@sesamiq.com>  Sat, 28 Aug 2010 00:43:55 +0800

macfanctld (0.0) lucid; urgency=low

  * Initial release

 -- Mikael Strom <mikael@sesamiq.com>  Fri, 27 Aug 2010 00:24:27 +0800



config.h
/*
 *  config.h -  Fan control daemon for MacBook
 *
 *  Copyright (C) 2010  Mikael Strom <mikael@sesamiq.com>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 */

#ifndef CONFIG_H_
#define CONFIG_H_

#define MAX_EXCLUDE	20
#define max(a,b)	(a > b ? a : b)
#define min(a,b)	(a < b ? a : b)

extern float temp_avg_floor;
extern float temp_avg_ceiling;

extern float temp_TC0P_floor;
extern float temp_TC0P_ceiling;

extern float temp_TG0P_floor;
extern float temp_TG0P_ceiling;

extern float fan_min;
extern float fan_max;

extern int log_level;

void read_cfg(char* name);

// array of sensors to exclude
extern int exclude[MAX_EXCLUDE];

#endif /* CONFIG_H_ */


control.c
/*
 *  control.c -  Fan control daemon for MacBook
 *
 *  Copyright(C) 2010  Mikael Strom <mikael@sesamiq.com>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <dirent.h>
#include <assert.h>
#include <unistd.h>
#include "config.h"

#define HWMON_DIR	"/sys/class/hwmon"
#define APPLESMC_ID	"applesmc"

// sensor control fan flags
#define CTL_NONE	0
#define CTL_AVG		1
#define CTL_TC0P	2
#define CTL_TG0P	3

#define N_DESC		(sizeof(sensor_desc) / sizeof(sensor_desc[0]))

#define SENSKEY_MAXLEN	16
#define SENSORS_MAX	100

struct {
	char *key;
	char *desc;
}

sensor_desc[] = {
	{"TB0T", "Battery TS_MAX Temp"},
	{"TB1T", "Battery TS1 Temp"},
	{"TB2T", "Battery TS2 Temp"},
	{"TB3T", "Battery Temp"},
	{"TC0D", "CPU 0 Die Temp"},
	{"TC0P", "CPU 0 Proximity Temp"},
	{"TG0D", "GPU Die - Digital"},
	{"TG0P", "GPU 0 Proximity Temp"},
	{"TG0T", "GPU 0 Die - Analog Temp"},
	{"TG0H", "Left Heat Pipe/Fin Stack Proximity Temp"},
	{"TG1H", "Left Heat Pipe/Fin Stack Proximity Temp"},
	{"TN0P", "MCP Proximity"},
	{"TN0D", "MCP Die"},
	{"Th2H", "Right Fin Stack Proximity Temp"},
	{"Tm0P", "Battery Charger Proximity Temp"},
	{"Ts0P", "Palm Rest Temp"}
};

struct sensor {
	int id;
	int excluded;
	char name[SENSKEY_MAXLEN];
	char fname[PATH_MAX];
	float value;
};

char base_path[PATH_MAX];
char fan1_out[PATH_MAX];
char fan2_out[PATH_MAX];
char fan1_man[PATH_MAX];
char fan2_man[PATH_MAX];

int sensor_count = 0;
int fan_count = 0;
float temp_avg = 0;
int fan_speed;

struct sensor *sensors = NULL;
struct sensor *sensor_TC0P = NULL;
struct sensor *sensor_TG0P = NULL;

// which sensor controls fan
int fan_ctl = 0;

// find and verify applesmc path in /sys/devices
void find_applesmc() {
	DIR *fd_dir;
	int ret;

	base_path[0] = 0;

	fd_dir = opendir(HWMON_DIR);

	if(fd_dir != NULL) {
		struct dirent *dir_entry;

		while((dir_entry = readdir(fd_dir)) != NULL && base_path[0] == 0) {
			if(dir_entry->d_name[0] != '.') {
				char name_path[PATH_MAX];
				int fd_name;

				sprintf(name_path, "%s/%s/device/name", HWMON_DIR, dir_entry->d_name);

				fd_name = open(name_path, O_RDONLY);

				if(fd_name > -1) {
					char name[sizeof(APPLESMC_ID)];

					ret = read(fd_name, name, sizeof(APPLESMC_ID) - 1);

					close(fd_name);

					if(ret == sizeof(APPLESMC_ID) - 1) {
						if(strncmp(name, APPLESMC_ID, sizeof(APPLESMC_ID) - 1) == 0) {
							char *dev_path;
							char *last_slash = strrchr(name_path, '/');

							if(last_slash != NULL) {
								*last_slash = 0;

								dev_path = realpath(name_path, NULL);

								if(dev_path != NULL) {
									strncpy(base_path, dev_path, sizeof(base_path) - 1);
									base_path[sizeof(base_path) - 1] = 0;
									free(dev_path);
								}
							}
						}
					}
				}
			}
		}
		closedir(fd_dir);
	}

	// create paths to fan and sensor
	if(base_path[0] == 0) {
		printf("Error: Can't find a applesmc device\n");
		exit(-1);
	}

	sprintf(fan1_out, "%s/fan1_output", base_path);
	sprintf(fan2_out, "%s/fan2_output", base_path);
	sprintf(fan1_man, "%s/fan1_manual", base_path);
	sprintf(fan2_man, "%s/fan2_manual", base_path);

	printf("Found applesmc at %s\n", base_path);
}

void read_sensors() {
	int i;
	for(i = 0; i < sensor_count; ++i) {
		if(! sensors[i].excluded) {
			// read temp value
			int fd = open(sensors[i].fname, O_RDONLY);
			if(fd < 0) {
				printf("Error: Can't open %s\n", sensors[i].fname);
				fflush(stdout);
			} else {
				char val_buf[16];
				int n = read(fd, val_buf, sizeof(val_buf));
				if(n < 1)
					printf("Error: Can't read  %s\n", sensors[i].fname);
				else
					sensors[i].value = (float)atoi(val_buf) / 1000.0;

				close(fd);
			}
		}
	}

	// calc average
	temp_avg = 0.0;
	int active_sensors = 0;

	for(i = 0; i < sensor_count; ++i) {
		if(! sensors[i].excluded) {
			temp_avg += sensors[i].value;
			++active_sensors;
		}
	}

	temp_avg = temp_avg / active_sensors;
}

void calc_fan() {
	fan_speed = fan_min;
	fan_ctl = CTL_NONE;

	// calc fan speed on average
	float fan_window = fan_max - fan_min;
	float temp_avg_window = temp_avg_ceiling - temp_avg_floor;
	float normalized_temp =(temp_avg - temp_avg_floor) / temp_avg_window;
	float fan_avg_speed =(normalized_temp * fan_window);

	if(fan_avg_speed > fan_speed) {
		fan_speed = fan_avg_speed;
		fan_ctl = CTL_AVG;
	}

	// calc fan speed for TC0P
	if(sensor_TC0P != NULL) {
		float temp_window = temp_TC0P_ceiling - temp_TC0P_floor;
		float normalized_temp =(sensor_TC0P->value - temp_TC0P_floor) / temp_window;
		float fan_TC0P_speed =(normalized_temp * fan_window);

		if(fan_TC0P_speed > fan_speed) {
			fan_speed = fan_TC0P_speed;
			fan_ctl = CTL_TC0P;
		}
	}

	// calc fan speed for TG0P
	if(sensor_TG0P != NULL) {
		float temp_window = temp_TG0P_ceiling - temp_TG0P_floor;
		float normalized_temp =(sensor_TG0P->value - temp_TG0P_floor) / temp_window;
		float fan_TG0P_speed =(normalized_temp * fan_window);

		if(fan_TG0P_speed > fan_speed) {
			fan_speed = fan_TG0P_speed;
			fan_ctl = CTL_TG0P;
		}
	}

	// finally clamp
	fan_speed = min(fan_max, fan_speed);
}

void set_fan() {
	char buf[16];

	// update fan 1
	int fd = open(fan1_out, O_WRONLY);
	if(fd < 0) {
		printf("Error: Can't open %s\n", fan1_out);
	} else {
		sprintf(buf, "%d", fan_speed);
		write(fd, buf, strlen(buf));
		close(fd);
	}

	// set fan 1 manual to zero
	fd = open(fan1_man, O_WRONLY);
	if(fd < 0) {
		printf("Error: Can't open %s\n", fan1_man);
	} else {
		strcpy(buf, "1");
		write(fd, buf, strlen(buf));
		close(fd);
	}

	// update fan 2
	if(fan_count > 1) {
		fd = open(fan2_out, O_WRONLY);
		if(fd < 0) {
			printf("Error: Can't open %s\n", fan2_out);
		} else {
			sprintf(buf, "%d", fan_speed);
			write(fd, buf, strlen(buf));
			close(fd);
		}

		// set fan 2 manual to zero
		fd = open(fan2_man, O_WRONLY);
		if(fd < 0) {
			printf("Error: Can't open %s\n", fan2_man);
		} else {
			strcpy(buf, "1");
			write(fd, buf, strlen(buf));
			close(fd);
		}
	}

	fflush(stdout);
}

void adjust() {
	read_sensors();
	calc_fan();
	set_fan();
}

void scan_sensors() {
	int i, j;
	struct stat buf;
	int result;

	sensor_TC0P = NULL;
	sensor_TG0P = NULL;

	// get number of fans
	result = stat(fan1_out, &buf);
	if(result != 0) {
		printf("No fans detected, terminating!\n");
		exit(-1);
	} else {
		fan_count = 1;
	}

	result = stat(fan2_out, &buf);
	if(result != 0) {
		printf("Found 1 fan.\n");
	} else {
		fan_count = 2;
		printf("Found 2 fans.\n");
	}

	// count number of sensors
	int count = 0;

	// more than 100 sensors is an error!
	while(count < SENSORS_MAX) {
		char fname[512];

		// sensor numbering start at 1
		sprintf(fname, "%s/temp%d_input", base_path, count + 1);
		result = stat(fname, &buf);

		if(result == 0)
			++count;
		else
			// done
			break;
	}

	sensor_count = count;

	// Get sensor id, labels and descriptions, check exclude list
	if(sensor_count > 0) {
		if(sensors != NULL)
			free(sensors);

		sensors = malloc(sizeof(struct sensor) * sensor_count);
		assert(sensors != NULL);

		printf("Found %d sensors:\n", sensor_count);

		for(i = 0; i < sensor_count; ++i) {
			char fname[512];

			// set id, check exclude list and save file name
			sensors[i].id = i + 1;
			sensors[i].excluded = 0;
			sprintf(sensors[i].fname, "%s/temp%d_input", base_path, sensors[i].id);

			for(j = 0; j < MAX_EXCLUDE && exclude[j] != 0; ++j) {
				if(exclude[j] == sensors[i].id) {
					sensors[i].excluded = 1;
					break;
				}
			}

			// read label
			sprintf(fname, "%s/temp%d_label", base_path, sensors[i].id);

			// set zero length
			sensors[i].name[0] = 0;

			FILE *fp = fopen(fname, "r");
			if(fp == NULL) {
				printf("Error: Can't open %s\n", fname);
			} else {
				char key_buf[SENSKEY_MAXLEN];
				memset(key_buf, 0, SENSKEY_MAXLEN);

				int n = fread(key_buf, 1, SENSKEY_MAXLEN - 1, fp);
				if(n < 1) {
					printf("Error: Can't read  %s\n", fname);
				} else {
					char *p_endl = strrchr(key_buf, '\n');

					// remove '\n'
					if(p_endl)
						*p_endl = 0;

					strncpy(sensors[i].name, key_buf, SENSKEY_MAXLEN);
				}
				fclose(fp);
			}
		}

		// for each label found
		for(i = 0; i < sensor_count; ++i) {
			if(! sensors[i].excluded) {
				// try to find TC0P and TG0P
				// if found, assign sensor_TC0P and sensor_TG0P for later use
				if(strcmp(sensors[i].name, "TC0P") == 0)
					sensor_TC0P = &sensors[i];
				else if(strcmp(sensors[i].name, "TG0P") == 0)
					sensor_TG0P = &sensors[i];
			}

			// print out sensor information.
			printf("\t%2d: ", sensors[i].id);

			// find in descriptions table
			int found = 0;
			for(j = 0; j < N_DESC && ! found; ++j)
				if(strcmp(sensors[i].name, sensor_desc[j].key) == 0) {
					found = 1;
					printf("%s - %s", sensor_desc[j].key, sensor_desc[j].desc);
				}

			if(! found)
				printf("%s - ?", sensors[i].name);

			printf(" %s\n", sensors[i].excluded ? "   ***EXCLUDED***" : "");
		}
	} else {
		printf("No sensors detected, terminating!\n");
		exit(-1);
	}

	fflush(stdout);
}

void logger() {
	int i;

	if(log_level > 0) {
		printf("Speed: %d, %sAVG: %.1fC" ,
		       fan_speed,
		       fan_ctl == CTL_AVG ? "*" : " ",
		       temp_avg);

		if(sensor_TC0P != NULL) {
			printf(", %sTC0P: %.1fC" ,
			       fan_ctl == CTL_TC0P ? "*" : " ",
			       sensor_TC0P->value);
		}

		if(sensor_TG0P != NULL) {
			printf(", %sTG0P: %.1fC" ,
			       fan_ctl == CTL_TG0P ? "*" : " ",
			       sensor_TG0P->value);
		}

		if(log_level > 1) {
			printf(", Sensors: ");
			for(i = 0; i < sensor_count; ++i)
				if(! sensors[i].excluded)
					printf("%s:%.0f ", sensors[i].name, sensors[i].value);
		}

		printf("\n");
		fflush(stdout);
	}
}


control.h
/*
 *  control.h -  Fan control daemon for MacBook
 *
 *  Copyright (C) 2010  Mikael Strom <mikael@sesamiq.com>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 */

#ifndef CONTROL_H_
#define CONTROL_H_

void find_applesmc();	// called once at startup, before anything else!
void scan_sensors();
void adjust();
void logger();
void deallocate_sensors(void);

#endif /* CONTROL_H_ */


macfanctl.c
/*
 *  macfanctl.c -  Fan control daemon for MacBook
 *
 *  Copyright (C) 2010  Mikael Strom <mikael@sesamiq.com>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  Note:
 *  This was written for MacBook Pro 5,1 and Ubutnu 10.04
 *  Requires applesmc-dkms
 *
 */

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <fcntl.h>
#include <signal.h>

#include "control.h"
#include "config.h"

#define PID_FILE	"/var/run/macfanctld.pid"
#define LOG_FILE	"/var/log/macfanctl.log"
#define CFG_FILE	"/etc/macfanctl.conf"

int running = 1;
int lock_fd = -1;
int reload = 0;

void signal_handler(int sig)
{
	switch (sig)
	{
	case SIGHUP:
		reload = 1;
		break;
	case SIGINT:
	case SIGTERM:
		running = 0;
		break;
	}
}

void daemonize() {
	// already a daemon
	if (getppid() == 1)
		return;

	// fork of new process
	pid_t pid = fork();

	// fork error
	if(pid < 0)
		exit(1);

	// parent exits
	if(pid > 0)
		exit(0);

	// child (daemon) continues

#ifdef DEBUG
	// time to attach debugger to this process
	sleep(20);
#endif
	// create a new session
	setsid();

	// set createfile permissions
#ifdef DEBUG
	umask(0);
#else
	umask(022);
#endif

	freopen(LOG_FILE, "w", stdout);
	freopen("/dev/null", "r", stdin);

	chdir("/");

	// create lockfile
	int lock_fd = open(PID_FILE, O_RDWR | O_CREAT, 0640);

	// open failed, we're a duplicate
	if(lock_fd < 0)
		exit(1);

	// lock failed - no idea what this means...
	if(lockf(lock_fd, F_TLOCK, 0) < 0)
		exit(0);

	// first instance continues...
	// write pid to file, and leave file open (blocking duplicates)
	char str[32];
	sprintf(str, "%d\n", getpid());
	write(lock_fd, str, strlen(str));
}


void usage() {
	printf("usage: macfanctld [-f]\n");
	printf("  -f  run in foregound\n");
	exit(-1);
}

int main(int argc, char *argv[]) {
	int i;
	int daemon = 1;

	// setup daemon
	signal(SIGCHLD, SIG_IGN);		// ignore child
	signal(SIGTSTP, SIG_IGN);		// ignore tty signals
	signal(SIGTTOU, SIG_IGN);
	signal(SIGTTIN, SIG_IGN);
	signal(SIGINT, signal_handler);		// catch Ctrl-C signal (terminating in foreground mode)
	signal(SIGHUP, signal_handler);		// catch hangup signal (reload config)
	signal(SIGTERM, signal_handler);	// catch kill signal

	for(i = 1; i < argc; ++i) {
		if(strcmp(argv[i], "-f") == 0)
			daemon = 0;
		else
			usage();
	}

	if(daemon)
		daemonize();
	else
		printf("Running in foreground, log to stdout.\n");

	// main loop setup
	read_cfg(CFG_FILE);

	find_applesmc();
	scan_sensors();

	running = 1;

	// main loop
	while(running) {
		adjust();

		logger();

		if(reload) {
			read_cfg(CFG_FILE);
			scan_sensors();
			reload = 0;
		}

		sleep(5);
	}

	// close pid file and delete it
	if(lock_fd != -1) {
		close(lock_fd);
		unlink(PID_FILE);
	}

	printf("Exiting.\n");

	return 0;
}


debian/init.d
#! /bin/sh
### BEGIN INIT INFO
# Provides:          macfanctl
# Required-Start:    $remote_fs
# Required-Stop:     $remote_fs
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: macfanctl initscript
# Description:       Start the macfanctl daemon
### END INIT INFO

DESC="Start the macfanctl daemon"
NAME=macfanctld
PATH=/sbin:/usr/sbin:/bin:/usr/bin
DAEMON=/usr/sbin/$NAME
PIDFILE=/var/run/$NAME.pid
SCRIPTNAME=/etc/init.d/$NAME
DAEMON_ARGS=""

# Exit if the package is not installed
[ -x "$DAEMON" ] || exit 0

# Load the VERBOSE setting and other rcS variables
. /lib/init/vars.sh

# Define LSB log_* functions.
# Depend on lsb-base (>= 3.0-6) to ensure that this file is present.
. /lib/lsb/init-functions

# Function that starts the daemon/service
do_start()
{
	start-stop-daemon --start --quiet --pidfile $PIDFILE --exec $DAEMON --test > /dev/null \
		|| return 1
	start-stop-daemon --start --quiet --pidfile $PIDFILE --exec $DAEMON -- \
		$DAEMON_ARGS \
		|| return 2
}

# Function that stops the daemon/service
do_stop()
{
	# Return
	start-stop-daemon --stop --quiet --retry=TERM/30/KILL/5 --pidfile $PIDFILE --name $NAME
	RETVAL="$?"
	[ "$RETVAL" = 2 ] && return 2
	start-stop-daemon --stop --quiet --oknodo --retry=0/30/KILL/5 --exec $DAEMON
	[ "$?" = 2 ] && return 2
	return "$RETVAL"
}

# Function that sends a SIGHUP to the daemon/service
do_reload() 
{
	start-stop-daemon --stop --signal 1 --quiet --pidfile $PIDFILE --name $NAME
	return 0
}

case "$1" in
  start)
	[ "$VERBOSE" != no ] && log_daemon_msg "Starting $DESC" "$NAME"
	do_start
	case "$?" in
		0|1) [ "$VERBOSE" != no ] && log_end_msg 0 ;;
		2) [ "$VERBOSE" != no ] && log_end_msg 1 ;;
	esac
	;;
  stop)
	[ "$VERBOSE" != no ] && log_daemon_msg "Stopping $DESC" "$NAME"
	do_stop
	case "$?" in
		0|1) [ "$VERBOSE" != no ] && log_end_msg 0 ;;
		2) [ "$VERBOSE" != no ] && log_end_msg 1 ;;
	esac
	;;
  status)
       status_of_proc "$DAEMON" "$NAME" && exit 0 || exit $?
       ;;
  restart|force-reload)
	log_daemon_msg "Restarting $DESC" "$NAME"
	do_stop
	case "$?" in
	  0|1)
		do_start
		case "$?" in
			0) log_end_msg 0 ;;
			1) log_end_msg 1 ;; # Old process is still running
			*) log_end_msg 1 ;; # Failed to start
		esac
		;;
	  *)
	  	# Failed to stop
		log_end_msg 1
		;;
	esac
	;;
  *)
	echo "Usage: $SCRIPTNAME {start|stop|status|restart|force-reload}" >&2
	exit 3
	;;
esac

:


config.c
/*
 *  config.c -  Fan control daemon for MacBook
 *
 *  Copyright (C) 2010  Mikael Strom <mikael@sesamiq.com>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include "config.h"

static FILE *fp;

// default values if no config file is found
float temp_avg_floor = 45;
float temp_avg_ceiling = 58;

float temp_TC0P_floor = 50;
float temp_TC0P_ceiling = 58;

float temp_TG0P_floor = 50;
float temp_TG0P_ceiling = 58;

float fan_min = 0;
float fan_max = 6500;

int log_level = 0;

// array of sensors to exclude
int exclude[MAX_EXCLUDE];

int match(char* name, char* buf) {
	char* start = buf;
	char* end = buf;

	if(strlen(buf) < 1)
		return 0;

	// skip preceeding ws
	while(*start && isblank(*start))
		++start;

	// skip to end
	while(*end)
		++end;

	// delete ws backwards from end
	while(end > start && isblank(*end)) {
		*end = 0;
		--end;
	}

	// compare strings
	return strcmp(name, start) == 0;
}

int get_val(char *buf) {
	while(isblank(*buf))
		++buf;

	if(! *buf)
		return -1;

	return atoi(buf);
}

// format is: name : integer
int read_param(char* name, int min_val, int max_val, int def) {
	fseek(fp, 0, SEEK_SET);

	while(1) {
		char buf[64];
		char *s = fgets(buf, sizeof(buf), fp);

		// exit when no more to read
		if(s == NULL)
			break;

		// skip comments
		if(buf[0] == '#' || buf[0] == '\n')
			continue;

		// find colon
		char *colon = strchr(buf, ':');
		if(colon == NULL) {
			printf("Ill formed line in config file: %s\n", buf);
			continue;
		}

		// terminate string at colon
		*colon = 0;

		if(match(name, buf)) {
			int val = get_val(colon + 1);	// get value

			if(val < 0) {
				printf("Ill formed line in config file: %s\n", buf);
				continue;
			} else {
				// clamp
				val = min(max_val, val);
				val = max(min_val, val);

				// success
				return val;
			}
		}
	}

	return def;
}

// format is: exclude : integer {integer}
void read_exclude_list() {
	fseek(fp, 0, SEEK_SET);

	while(1) {
		char buf[256];
		char *s = fgets(buf, sizeof(buf), fp);

		// exit when no more to read
		if(s == NULL)
			break;

		// skip comments
		if(buf[0] == '#' || buf[0] == '\n')
			continue;

		// find colon
		char *colon = strchr(buf, ':');
		if(colon == NULL) {
			printf("Ill formed line in config file: %s\n", buf);
			continue;
		}

		// terminate string at colon
		*colon = 0;

		if(match("exclude", buf)) {
			int i;
			char* values = colon + 1;

			// get values
			for(i = 0; i < MAX_EXCLUDE; ++i) {
				while(isspace(*values) || *values == ',')
					++values;

				if(isdigit(*values)) {
					int val = get_val(values);
					
					if(val < 0) {
						printf("Ill formed line in config file: %s\n", buf);
						return;
					}
					
					exclude[i] = val;
					
					while(isdigit(*values))
						++values;
				} else if(*values == 0) {
					// done
					return;
				} else {
					printf("Ill formed line in config file: %s\n", buf);
					return;
				}
			}
		}
	}
}
 
void read_cfg(char* name) {
	memset(exclude, 0, sizeof(exclude));

	fp = fopen(name, "r");

	if(fp != NULL) {
		temp_avg_ceiling = read_param("temp_avg_ceiling",	0, 90, 50);
		temp_avg_floor = read_param("temp_avg_floor",		0, temp_avg_ceiling - 1, 40);

		temp_TC0P_ceiling = read_param("temp_TC0P_ceiling",	0, 90, 65);
		temp_TC0P_floor = read_param("temp_TC0P_floor",		0, temp_TC0P_ceiling - 1, 50);

		temp_TG0P_ceiling = read_param("temp_TG0P_ceiling",	0, 90, 80);
		temp_TG0P_floor = read_param("temp_TG0P_floor",		0, temp_TG0P_ceiling - 1, 65);

		fan_min = read_param("fan_min", 0, 6200, 0);

		log_level = read_param("log_level", 0, 2, 0);
		
		read_exclude_list();

		fclose(fp);
	} else {
		printf("Could not open config file %s\n", name);
	}

	printf("Using parameters:\n");

	printf("\ttemp_avg_floor: %.0f\n", temp_avg_floor);
	printf("\ttemp_avg_ceiling: %.0f\n", temp_avg_ceiling);

	printf("\ttemp_TC0P_floor: %.0f\n", temp_TC0P_floor);
	printf("\ttemp_TC0P_ceiling: %.0f\n", temp_TC0P_ceiling);

	printf("\ttemp_TG0P_floor: %.0f\n", temp_TG0P_floor);
	printf("\ttemp_TG0P_ceiling: %.0f\n", temp_TG0P_ceiling);

	printf("\tfan_min: %.0f\n", fan_min);

	if(exclude[0] != 0) {
		int i;

		printf("\texclude: ");

		for(i = 0; i < MAX_EXCLUDE && exclude[i] != 0; ++i)
			printf("temp%d_input ", exclude[i]);
		printf("\n");
	}

	printf("\tlog_level: %d\n", log_level);
}
